aten/src/ATen/native/cuda/TensorTransformations.cu

```diff
@@ -126,4 +127,65 @@ Tensor flip_cuda(const Tensor& self, IntList dims) {
   return out_tensor;
 }
 
+template <typename scalar_t>
+__global__
+void roll_cuda_kernel(scalar_t* in_tensor, scalar_t* out_tensor, int64_t N,
+                      int64_t roll_dim, int64_t shift, int64_t start,
+                      int64_t size, int64_t stride, int64_t total_dims) {
+  int64_t linear_index = blockIdx.x * blockDim.x + threadIdx.x;
+  if (linear_index >= N) {
+    return;
+  }
+  // roll dim idx is the index of linear_index along the rolling dimension.
+  int64_t roll_dim_idx = linear_index % (stride * size) / stride;
```

Comment:
the RHS of the % can be 0 here, which according to the C++ standard is undefined behavior; stride can be 0 if you have an expanded tensor, e.g.:

```
torch.randn(3,1,device='cuda').expand(3,4).stride()
(1, 0)
```

Reply:
oh actually, this isn't true because you make it contiguous, but it will break once that assumption is removed.

Reply:
adding example of what would fail if not made contiguous, for future reference:

>>> x = torch.randn(3,1)
>>> x
tensor([[-0.0317],
        [ 0.7835],
        [-0.5729]])
>>> x.stride()
(1, 1)
>>> x.expand(3,4)
tensor([[-0.0317, -0.0317, -0.0317, -0.0317],
        [ 0.7835,  0.7835,  0.7835,  0.7835],
        [-0.5729, -0.5729, -0.5729, -0.5729]])
>>> x.expand(3,4).stride()
(1, 0)
>>> x.expand(3,4).contiguous().stride()
(4, 1)